src/platforms/esp32/components/avm_builtins/gpio_driver.c

```diff
@@ -282,6 +287,7 @@ static term gpiodriver_close(Context *ctx)
         }
     }
 
+    ctx->platform_data = NULL;
     globalcontext_unregister_process(glb, gpio_atom_index);
     free(gpio_data);
```

Comment:
Is free-ing platform_data and nullifying the reference supposed to be done here or should it be deferred to `context_destroy`.  Perhaps @pguyot should weigh in on best practices here.

Reply:
I would like to hear what @pguyot has to say. This is also what is done in the I2C driver, nullifying the reference was very recently added, which pointed me in the direction of using this, and does fix the crash described in #816.

Reply:
I believe this change is correct but unsufficient. If `platform_data` is freed (as it currently is in `gpiodriver_close`), it should be nullified as indeed `context_destroy` will attempt to free it again if it is not null, hence crash #816.

The reason the free was added in `context_destroy` is to avoid any memory leak. And free is called at the end of the destroy function, after the process is made unreachable (is removed from the process table), to cope with cases where the platform data is called from interruptions. This is typically the "ugly hack" mentioned in i2c driver.

It looks like there still is a problem with GPIO driver, though. `gpiodriver_close` loops over listeners like `gpiodriver_remove_int`, but doesn't do it as well. It does call `gpio_isr_handler_remove`, but it doesn't call `sys_unregister_listener` unlike `gpiodriver_remove_int`.

Reply:
Indeed. I was unsafely removing the `listener.listeners_list_head` manually (without the lock), rather than using the safe `sys_unregister_listener`

Reply:
This is fixed.

Reply:
We have 9 duplicate lines. Do you think we should factorize them?
277-286 and 457-466.

Reply:
Refactoring may be a good idea, it could alleviate future discrepancies between the `gpiodriver_close` and `gpiodriver_remove_int` handling of removing the listener.

If I were to replace the duplicate code with a common `remove_gpio_listener` function would it make sense to make that an inline function in this case?

Reply:
Of course the factorized code can be inlined.